doc/reference/configuration/configuration_reference.rst

```diff
@@ -2877,6 +2878,120 @@ This section describes configuration parameters related to roles.
+
+    There are two types of application roles:
+
+    -   Built-in roles. For example, the `CRUD <https://github.com/tarantool/crud>`__ module provides the ``roles.crud-storage`` and ``roles.crud-router`` roles that enable CRUD operations in a sharded cluster.
+    -   Custom roles. For example, you create a custom role to define a stored procedure or implement a supplementary service, such as an e-mail notifier or a replicator.
+
+    To create a custom role, you need to define the following methods and fields:
+
+    -   ``validate()``
+
+        Validate a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``validate()`` should throw an error if the validation fails.
+
+    -   ``apply()``
+
+        Apply a role's configuration.
+        This method is called when the initial configuration is applied or the configuration is reloaded for the instance with this role.
+        ``apply()`` should throw an error if the specified configuration can't be applied.
+
+    -   ``stop()``
+
+        Stop a role.
+        This method is called on configuration reload if the role is removed from ``roles`` for the given instance.
+        ``stop()`` should cancel all background fibers and clean up used resources.
+        If these actions can't be performed, ``stop()`` should throw an error.
+
+    -   (Optional) ``dependencies``
+
+        Define a role's dependencies as an array.
+        For roles that depend on each other, their ``validate()`` and ``apply()`` methods are executed taking into account the dependencies.
+        Suppose, there are three roles that depend on each other as follows:
+
+        ..  code-block:: none
+
+            role1
+                └─── role2
+                         └─── role3
+
+        In this case, ``validate()`` and ``apply()`` for these roles are executed in the following order:
+
+        ..  code-block:: none
+
+            role3 -> role2 -> role1
+
+
+    **Example: Built-in role**
+
+    The examples below show how to enable the router and storage-specific roles provided by the `CRUD <https://github.com/tarantool/crud>`__ module:
+
+    -   ``roles.crud-router``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: routers
+            :end-at: roles.crud-router
+            :dedent:
+
+    -   ``roles.crud-storage``:
+
+        ..  literalinclude:: /code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud/config.yaml
+            :language: yaml
+            :start-at: storages
+            :end-at: roles.crud-storage
+            :dedent:
+
+    You can find the full example here: `sharded_cluster_crud <https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets/sharding/instances.enabled/sharded_cluster_crud>`_.
+
+    **Example: Custom role**
```

Comment:
For me, it feels beyond the configuration reference scope. Can we create a separate 'how to write a role' document and link it from here?

Reply:
As a temporary workaround, this example will be placed in the configuration reference but outside the `roles/roles_cfg` descriptions. As some moment, all such content will be moved to the `Developing applications` section (https://github.com/tarantool/doc/issues/3920).